### PR TITLE
Fix #187 replace circular list to enable native compilation

### LINKED ISSUE #187: Get loop list warning when doing native compilation
I'm trying to native compile `powerline.el` on `Emacs 28.0.50`, this warning
```
/path/to/elpa/powerline-20210313.1621/powerline.el: Error: List contains a loop ("22", . #0)
```
makes `powerline` failed to compile.

Currently I add regexp `powerline` into `comp-deferred-compilation-deny-list` to walk around this issue.
```emacs-lisp
(setq comp-deferred-compilation-deny-list '("powerline"))
```

### PR DESCRIPTION
**Pull request by wyuenho**
Replace the circular list in pl/pattern with a ring so the new and shiny JIT can compile powerline. Added some minor memory optimizations too.

### CODE REVIEW

**Comment by wyuenho** [comment]
@milkypostman mind taking a look? Should be relatively straightforward.

File: powerline-separators.el [hunk 1]
```diff
@@ -79,8 +80,10 @@ RED, GREEN and BLUE should be between 0.0 and 1.0, inclusive."
 (defun pl/pattern (lst)
   "Turn LST into an infinite pattern."
   (when lst
-    (let ((pattern (cl-copy-list lst)))
-      (setcdr (last pattern) pattern))))
+    (let ((r (make-ring (length lst))))
```
Comment (milkypostman) [inline powerline-separators.el:83]: could we use `ring-convert-sequence-to-ring`?

**Review by milkypostman — CHANGES_REQUESTED** [review]

File: powerline-separators.el (hunk 1 above)
Comment (wyuenho) [reply powerline-separators.el]: Didn't even know it existed lol

**Review by milkypostman — APPROVED** [review]

**Comment by milkypostman** [comment]
Thanks for your work on this @wyuenho !

**Pull request merged**